### ISSUE TITLE
SC1-4-1-inline-links-distinguishable

### ISSUE THREAD

**Issue by JenniferChadwick** [act-rules/act-rules.github.io#92]
name: Inline links are distinguishable

description: |
This rule checks that links that are embedded in a block of text and where color alone is used to identify them, have a sufficient color contrast ratio with the surrounding text, and additional cues are given when the link receives focus or hover. 

success_criterion:

- 1.4.1 # Use of Color

test_aspects:

- Dom Tree 
- CSS Styling

authors:
- Wilco Fiers
- Jennifer Chadwick
- Anne Thyme Nørregaard

### Applicability

The rule applies to any HTML or SVG element with the [semantic role][] of link that lives up to the following criteria:
* is [focusable][]
* contains [visible text][] that doesn't have a [distinguishing style][], a [distinguishing border][], a [distinguishing box-shadow][] or a background-image.
* is in a [block of text][] that has [visible text][], which 
** is NOT contained in any link, and 
** does not exclusively consist of [separator characters][].

### Expectation 1

Within the same [block of text] the [contrast ratio](https://www.w3.org/TR/WCAG20/#contrast-ratiodef) is 3:1 or greater between either 
- the [color][] of the [visible text][] of the link and [adjacent][] [visible text], or 
- the [background color][] of the link and the background color of any adjacent visible text.

### Expectation 2

When the link's [focus state is triggered][], the link has a [distinguishing style][], a [distinguishing border][], a [distinguishing box-shadow][] or a background-image.

### Expectation 3

When the link's [hover state is triggered][], the link has a [distinguishing style][], a [distinguishing border][], a [distinguishing box-shadow][] or a background-image.

## Assumptions

- This tests assumes that [distinguishing style][], [distinguishing border][] or one of the following CSS properties  is used to make the link visually evident: color, background-color or background-image. 
- This test assumes that the 3:1 [contrast ratio](https://www.w3.org/TR/WCAG20/#contrast-ratiodef) between link text and surrounding text is sufficient to meet WCAG 2.0. This value is part of technique G183, but is not specified in the 1.4.1 success criterion.
- The methods described in [distinguishing style][] is assumed to be sufficiently distinguishable.
- The methods described in [distinguishing border][] is assumed to be sufficiently distinguishable.
- The methods described in [distinguishing box-shadow][] is assumed to be sufficiently distinguishable.

## Accessibility Support

There are no major accessibility support issues known for this rule.

## Background

- [G183: Using a contrast ratio of 3:1 with surrounding text and providing additional visual cues on focus for links or controls where color alone is used to identify them](http://www.w3.org/TR/WCAG20-TECHS/G183.html) 
- [F73: Failure of Success Criterion 1.4.1 due to creating links that are not visually evident without color vision](https://www.w3.org/TR/2014/NOTE-WCAG20-TECHS-20140311/F73.html)
- [C15: Using CSS to change the presentation of a user interface component when it receives focus](https://www.w3.org/TR/WCAG20-TECHS/C15.html) 

## Test Cases

@@@ Test cases: breaks, margins, content in lots of spans etc. 

### Passed

```html
<!-- Link uses the browser's default link styling -->
<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>
```

```html
<!-- Blue link in block of black text with text-decoration none has underline on hover and focus -->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
}

a:hover, a:focus {
    text-decoration: underline;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>

</body>
```

```html
<!-- Block of text only consist of one letter that is not contained within a link -->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
}

a:hover, a:focus {
    text-decoration: underline;
}
</style>
</head>

<body>

<p>A<a href="http://w3.org/WAI"> link text</a></p>

</body>
```

```html
<!-- -->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
}

a:hover, a:focus {
    text-decoration: underline;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>

</body>
```

```html
<!-- Additional cues on hover/focus is change in font-family-->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
    font-family: "Times New Roman", Times, serif;
}

a:hover, a:focus {
    font-family: Arial, Helvetica, sans-serif;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>

</body>
```

```html
<!-- Additional cues on hover/focus is change in font-weight -->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
    font-weight: normal;
}

a:hover, a:focus {
    font-weight: bold;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>

</body>
```

```html
<!-- Additional cues on hover/focus is italics -->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
    font-style: normal;
}

a:hover, a:focus {
    font-style: italic;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>

</body>
```

```html
<!-- Additional cues on hover/focus is change of background-color -->
<head>
<style>
p {
    color: black;
    background-color: white;
}

a:link {
    color: blue;
    text-decoration: none;
}

a:hover, a:focus {
    background-color: red;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>

</body>
```

```html
<!-- Additional cues on hover/focus is background-image -->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
}

a:hover, a:focus {
    background-image: url("arrow.gif");
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>

</body>
```

```html
<!-- Button that is styled to look like a link and has role of link -->
<head>
<style>
p {
    color: black;
}

button {
     background:none;
     color:blue;
     border:none; 
     padding:0;
     font: inherit;
     cursor: pointer;
}

button:hover, button:focus {
    text-decoration: underline;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <button role="link">button that acts like a link</button>.</p>

</body>
```

´´´html

´´´

´´´html

´´´
### Failed

@@@ split out <a> and a: link underlined
```html
<!-- Too low contrast between link text and surrounding text-->

<head>
<style>
p {
    color: red;
}

a:link {
    color: blue;
    text-decoration: none;
}

a:hover, a:focus {
    text-decoration: underline;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>

</body>
```

```html
<!-- No additional cues on hover  -->

<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
}

a:focus {
    text-decoration: underline;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>

</body>
```

```html
<!-- Additional cues on focus -->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
}

a:hover {
    text-decoration: underline;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>

</body>
```

´´´html

´´´

### Inapplicable 

```html
<!-- -->
<!-- <a> element doesn't have href="", so it doesn't have role of link -->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a>a element without href attribute</a>.</p>

</body>
```

```html
<!-- <a> element has had it's role changed -->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI" role="button">link that has had its role changed</a>.</p>

</body>
```

```html
<!-- No visible text in the block of text the link is contained within -->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
}
</style>
</head>

<body>

<p><a href="http://w3.org/WAI">Read more</a><span style="display:none">about our latest organisational changes.</span></p>

</body>
```

```html
<!-- Breadcrumb with only links and separator characters -->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
}

</style>
</head>

<body>

<p>... / <a>About us</a> / <a>CSR</a> / <a>Accessibility</a></p>

</body>
```

```html
<!-- -->
<head>
<style>
p {
    color: black;
}

a:link {
    color: blue;
    text-decoration: none;
}
</style>
</head>

<body>

<p>This is a block of text<a href="http://w3.org/WAI"></a>.</p>

</body>
```

```html
<!-- Link is underlined, therefore this rule is not applicable -->
<head>
<style>
p {
    color: black;
}

a:link {
    text-decoration: underline;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>

</body>
```

```html
<!-- -->
<head>
<style>
a:link {
    text-decoration: none;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <span style="text-decoration: underline;"><a href="http://w3.org/WAI">link text that is underlined</a><span>.</p>

</body>
```

```html
<!-- -->
<head>
<style>
a:link {
    text-decoration: none;
}
</style>
</head>

<body>

<p><a href="http://w3.org/WAI">link 1</a> | <a href="http://w3.org/WAI">link 2</a>.</p>

</body>
```

```html
<!-- -->
<head>
<style>
a:link {
    text-decoration: none;
}
</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI"><span style="text-decoration: underline;">link text that is underlined<span></a>.</p>

</body>
```

```html
<!-- -->
<!-- button that looks like a link but has the role of button -->
<head>
<style>
p {
    color: black;
}

button {
     background:none;
     color:blue;
     border:none; 
     padding:0;
     font: inherit;
     cursor: pointer;
}

button:hover {
    text-decoration: underline;
}

button:focus {
    text-decoration: underline;
}

</style>
</head>

<body>

<p>This is a block of text with an embedded <button>button that acts like a link</button>.</p>

</body>
```

```html
<!-- Link text color has too low contrast with background, but this is a violation of success criteria 1.4.3 Contrast (Minimum) and 1.4.6 Contrast (Enhanced), not 1.4.1 Use of Color  -->
<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>
```

```html
<!-- Blue link in block of black text with text-decoration none has underline on hover and focus -->
<head>
<style>
p {
    color: black;
    background-color: white;
}

a:link {
    color: yellow;
    text-decoration: none;
}

</style>
</head>

<body>

<p>This is a block of text with an embedded <a href="http://w3.org/WAI">link text</a>.</p>

</body>
```

@@@ add a:focus test cases for Passed test cases
@@@ fix up text color in first failed test case
@@@ comments in test cases about why they are failing

[visible on the page]: ../pages/algorithms/visible-on-page.html
[color]: ../pages/algorithms/color.html
[focusable]: ../pages/algorithms/focusable.html
[non-empty]: ../pages/algorithms/non-empty.html
[rendered text]: ../pages/algorithms/rendered-text.html
[separator characters]: ../pages/algorithms/separator-characters.html
[adjacent]: ../pages/algorithms/adjacent.html
[phrasing content]: ../pages/algorithms/phrasis-content.html
[nearest]: ../pages/algorithms/nearest.html
[block-like]: ../pages/algorithms/block-like.html
[ancestor]: ../pages/algorithms/ancestor.html
[background color]: ../pages/algorithms/background-color.html
[seperator characters]: ../pages/algorithms/seperator-characters.html
[adjacent]: ../pages/algorithms/adjacent.html
[block of text]: ../pages/algorithms/block-of-text.html
[distinguishing style]: ../pages/algorithms/distinguishing-style.html
[distinguishing border]: ../pages/algorithms/distinguishing-border.html
[semantic role]: ../pages/algorithms/semantic-role.html
[hover state is triggered]: ../pages/algorithms/hover-state-is-triggered.html
[focus state is triggered]: ../pages/algorithms/focus-state-is-triggered.html
[visible text]: ../pages/algorithms/visible-text.html
[distinguishing box-shadow]: ../pages/algorithms/distinguishing-box-shadow.html

-------

# Definitions

## Block of text

A block of text is a run of phrasing content containing no consecutive line breaks.

Contained within one [block-like][] [ancestor][]
has [rendered text][] or [phrasing content](https://www.w3.org/TR/html/dom.html#phrasing-content) with rendered text

If more than one `br` element is used, followed by additional [rendered text][] or [phrasing content](https://www.w3.org/TR/html/dom.html#phrasing-content) with rendered text, this is considered a new block of text.

What if `pre`element is used?

[block-like]: ../pages/algorithms/block-like.html
[rendered text]: ../pages/algorithms/rendered-text.html
[ancestor]: ../pages/algorithms/ancestor.html

## Distinguishing Border

Look at the width, style, and color of the border-top, border-bottom or outline.

For one of these to be a distinguishing border, all the following must be true:

* The *-width is greater than 0
* The *-style is not set to none 
* The *-color is not transparent

If either border-top, border-bottom or outline is distinguishable, the element has a distinguishing border.

### Assumptions
* Use of a border, of 1 or more pixels, not set to none, and not with a color of transparent, is assumed to be sufficiently distinguishable.

## Distinguishing style

Compare the style properties font-family, font-weight, font-style, text-decoration of the element to that of the surrounding text.

If any of the properties have a different value for the one than for the other, the style is distinguishable.

### Assumptions
- This definition assumes that one of the following shorthand CSS properties (or the related expanded properties e.g. font-family) is used to make the element visually evident: font or text-decoration
- This definition assumes that any change in font is sufficiently distinguishable, and that fonts are loaded when they are present

**Comment by Brynanders** [comment]
Should ‘relative luminance’ be under ‘Assumptions’ for ‘contrast ratio’ instead of as a separate sentence under Expectations 1?

**Comment by JenniferChadwick** [comment]
Thanks, @Brynanders I can agree.   The relative luminance is a measurement that in the past has been included as part of the color contrast measurement, so I put it in Expectations to ensure it would be tested.  However,  we could place it under Assumptions - assuming that the tester knows this or will be including luminance as part of their colour contrast check.

**Comment by WilcoFiers** [comment]
This rule may seem obvious, but it I don't think it is. Here are some questions to answer:

- What is a text link? Any `a[href]` with text in it? Any `role="link"`?

- How can you tell that "color is being used to distinguish"? Which (CSS?) properties should or should not be used, or should be different?

- What is "surrounding text"? When is text surrounding, and when is it not? Does it just mean adjecent to? If so, how much space is supposed to be in between before it isn't "surrounding". Does that include text above and below it, or just left and right from it? What about when there is only text to the left OR to the right, is it still "surrounded"?

Have a look at the existing draft. Would it make sense to try and update it instead of a rewrite?
https://auto-wcag.github.io/auto-wcag/drafts/SC1-4-1-link-style-in-text.html

**Comment by JenniferChadwick** [comment]
@WilcoFiers - thank you for the questions.  They seem to be answered in existing draft, so I attempted (shamelessly) to pull the Steps from the draft into this rule to add those additional details.   I'm struggling to map the draft to the new rule template and I don't want to lose important information. 
I'd appreciate some help from @skotkjerra, thanks!

**Comment by Brynanders** [comment]
Made a load of changes based on comments including feedback from @skotkjerra, @kasperisager, @WilcoFiers including changes to:
- Description
- Test aspect
- Authors
- Applicability
- Expectations

**Comment by annethyme** [comment]
Should rule name be inline-links-distinguishable since it's about much more than just color contrast?

**Comment by gianwild** [comment]
+1

Cheers,
Gian

**Comment by kasperisager** [comment]
This bit from __Applicability__...

> - The color contrast between either the [color]·[] of the link and any adjacent phrasing content within its nearest block-like ancestor or between the background-color of the link and any adjacent phrasing content within its nearest block-like ancestor is more than 3:1.

...should be moved to __Expectation 1__.

**Comment by kasperisager** [comment]
Regarding the "is focusable" point in the applicability section, one could argue that we also need to cover cases where people have applied bad hover styles to semantic links that aren't focusable such `<span role="link">...</span>`. I'm however leaning towards leaving out those cases and instead implement a separate rule that ensures that [widget roles](https://www.w3.org/TR/wai-aria/#widget_roles) can actually receive focus.

**Comment by Brynanders** [comment]
I am not sure about the 4th Pass test case- the only distinguishing feature on hover is a subtle change from Aria to a Times font. Can font-family alone be used as a distinguishing style? to pass Expectations 3? For the record, I am fine with i,t especially as to change it would require a very subjective manual review to pass the check... but as a visually impaired user I do find the change somewhat subtle so thought I might point this out https://www.screencast.com/t/B0fgdSEKjY.

**Comment by annethyme** [comment]
Suggested changes for "The link’s nearest block-like ancestor has non-empty rendered text that is NOT contained in any link, and which does not exclusively consist of separator characters (in short: the link is in a block of text)." in Applicability (from meeting between Siteimprove and Deque on May 7): 

- What about line breaks between text and link, but in same paragraph? (inapplicable test cases should cover this as well)
- "link has to be inline text and has to have inline text before or after it that is not itself a link or separator characters."
- "link is a part of text run contents that is not itself a link or separator characters"
- look at CSS spec (CSS display model)

**Comment by kasperisager** [comment]
I think starting off with a robust definition of "block of text" would be a good start. Something along the lines of: A block of text is a run of [phrasing content](http://w3c.github.io/html/dom.html#phrasing-content) containing no consecutive [line breaks](http://w3c.github.io/html/textlevel-semantics.html#elementdef-br).

**Comment by annethyme** [comment]
Changing Applicability due to conversation on https://github.com/auto-wcag/auto-wcag/issues/131.

### Old Applicability: 

The rule applies to any HTML or SVG element with the [semantic role][] of link that lives up to the following criteria:
- is [focusable][]
- is [visible on the page][]
- contains [rendered text][]
- is in a [block of text][] that has [rendered text][] that is NOT contained in any link, and which does not exclusively consist of [separator characters][]
- The link (or another element that contains all that link’s content) does not have a [distinguishing style](https://auto-wcag.github.io/auto-wcag/drafts/SC1-4-1-link-style-in-text.html#distinguishing-style), a [distinguishing border](https://auto-wcag.github.io/auto-wcag/drafts/SC1-4-1-link-style-in-text.html#distinguishing-border), or a background-image.

### Updated Applicability:

The rule applies to any HTML or SVG element with the [semantic role][] of link that lives up to the following criteria:
- is [focusable][]
- contains [rendered text][] that is [visible on the page][]
- is in a [block of text][] that has [rendered text][] that is NOT contained in any link, and which does not exclusively consist of [separator characters][]
- The link (or another element that contains all that link’s content) does not have a [distinguishing style](https://auto-wcag.github.io/auto-wcag/drafts/SC1-4-1-link-style-in-text.html#distinguishing-style), a [distinguishing border](https://auto-wcag.github.io/auto-wcag/drafts/SC1-4-1-link-style-in-text.html#distinguishing-border), or a background-image.

**Comment by annethyme** [comment]
Should a manual review for background image distinghuishability be related to this success criterion or https://www.w3.org/TR/WCAG21/#non-text-contrast ???

**Comment by kooleo** [comment]
does this apply only for links that are part of a paragraph or also for list of links under a heading. For example 

Fruits 
- Apple 
- Banana
- Grapes 

Should there be a contrast of 3:1 between heading "Fruits" and link text "Apple".

**Comment by JenniferChadwick** [comment]
This rule is currently on hold as of Sept 17th.

**Comment by jkodu** [comment]
@JenniferChadwick 

Closing this issue due to inactivity. Feel free to open it to restart discussion if necessary.
